e2etest/bgptests/bgp.go

```diff
@@ -1718,6 +1718,290 @@ var _ = ginkgo.Describe("BGP", func() {
+						},
+					},
+					{
+						ObjectMeta: metav1.ObjectMeta{Name: "bgp-test-3"},
+						Spec: metallbv1beta1.IPAddressPoolSpec{
+							Addresses: []string{
+								"192.168.20.0/32",
+							},
+						},
+					},
+				},
+				Peers:   metallb.PeersForContainers(FRRContainers, ipfamily.IPv4),
+				BGPAdvs: []metallbv1beta1.BGPAdvertisement{emptyBGPAdvertisement},
+			}
+
+			for _, c := range FRRContainers {
+				err := frrcontainer.PairWithNodes(cs, c, ipfamily.IPv4)
+				framework.ExpectNoError(err)
+			}
+
+			err := ConfigUpdater.Update(resources)
+			framework.ExpectNoError(err)
+
+			for _, c := range FRRContainers {
+				validateFRRPeeredWithAllNodes(cs, c, ipfamily.IPv4)
+			}
+
+			svcTweakWithLabel := func(svc *corev1.Service) {
+				if svc.Labels == nil {
+					svc.Labels = make(map[string]string)
+				}
+				svc.Labels["test"] = "e2e"
+			}
+
+			svc1, _ := testservice.CreateWithBackend(cs, f.Namespace.Name, "external-local-lb-1", svcTweakWithLabel)
+			svc2, _ := testservice.CreateWithBackend(cs, f.Namespace.Name, "external-local-lb-2", svcTweakWithLabel)
+			svc3, _ := testservice.CreateWithBackend(cs, f.Namespace.Name, "external-local-lb-3", svcTweakWithLabel)
+			defer func() {
+				testservice.Delete(cs, svc1)
+				testservice.Delete(cs, svc2)
+				testservice.Delete(cs, svc3)
+			}()
+
+			framework.ExpectEqual(f.Namespace.Name, f.Namespace.Name)
+			for _, i := range svc1.Status.LoadBalancer.Ingress {
+				ginkgo.By("validate LoadBalancer IP is allocated from 1st higher priority address pool")
+				ingressIP := e2eservice.GetIngressPoint(&i)
+				framework.ExpectEqual(ingressIP, "192.168.10.0")
+			}
+			for _, i := range svc2.Status.LoadBalancer.Ingress {
+				ginkgo.By("validate LoadBalancer IP is allocated from 2nd higher priority address pool")
+				ingressIP := e2eservice.GetIngressPoint(&i)
+				framework.ExpectEqual(ingressIP, "192.168.5.0")
+			}
+			for _, i := range svc3.Status.LoadBalancer.Ingress {
+				ginkgo.By("validate LoadBalancer IP is allocated from default address pool")
+				ingressIP := e2eservice.GetIngressPoint(&i)
+				framework.ExpectEqual(ingressIP, "192.168.20.0")
+			}
+
+			allNodes, err := cs.CoreV1().Nodes().List(context.Background(), metav1.ListOptions{})
+			framework.ExpectNoError(err)
+
+			for _, c := range FRRContainers {
+				validateService(cs, svc1, allNodes.Items, c)
+				validateService(cs, svc2, allNodes.Items, c)
+				validateService(cs, svc3, allNodes.Items, c)
+			}
+		})
+	})
+
+	ginkgo.Context("IPV4 - Validate service allocation in address pools", func() {
+		ginkgo.AfterEach(func() {
+			// Clean previous configuration.
+			err := ConfigUpdater.Clean()
+			framework.ExpectNoError(err)
+		})
+		ginkgo.It("with namespace and service label", func() {
+			resources := metallbconfig.ClusterResources{
+				Pools: []metallbv1beta1.IPAddressPool{
+					{
+						ObjectMeta: metav1.ObjectMeta{Name: "bgp-test-1"},
```

Comment:
This is making me think: what happens if we have a pool with a namespace selector and a label selector? They are or-ed right?